src/evaluate.py

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""
+This module provides functionalities for evaluating the YAMLsmith model designed to
+generate YAML state configurations based on naturural language input. It includes
+functions to load models and testing data, preprocess data, compute evaluation
+metrics (Exact Match, Schema Correctness, and Levenshtein Distance), and record the
+evaluation results into a CSV file.
+
+The evaluation process involves loading a trained model and a dataset, generating
+answers for the provided questions, comparing these answers against expected outputs,
+computing specified metrics, and finally writing these metrics to a CSV file for
+further analysis.
+
+Usage:
+- This module is intended to be run as a script. The main function orchestrates
+  loading of data and model, evaluation of the model, and writing out results.
+"""
+
+# pylint: disable=import-error
+import csv
+import yaml
+import Levenshtein
+from transformers import AutoModelForCausalLM, GPT2Tokenizer
+from utils import read_and_preprocess_data, load_yaml_dataset
+
+
+def load_model(model_path, model_type="YAMLsmith"):
+    if model_type == "YAMLsmith":
+        evaluation_tokenizer = GPT2Tokenizer.from_pretrained(model_path)
+        trained_model = AutoModelForCausalLM.from_pretrained(model_path)
+        evaluation_tokenizer.pad_token = evaluation_tokenizer.eos_token
+        return evaluation_tokenizer, trained_model
+    # TO DO: load other models like Ansible Lightspeed
+    return None, None
+
+
+def exact_match_score(expected_answer, answer):
+    """Calculate the Exact Match score between an expected and a provided answer.
+
+    Parameters:
+        expected_answer (str): The correct answer.
+        answer (str): The model's predicted answer.
+
+    Returns:
+        int: 1 if the answers match exactly, 0 otherwise.
+    """
+    return 100 if expected_answer == answer else 0
+
+
+def schema_correct(answer):
```

Comment:
IMHO this should be YAML correct, the schema would be the result of whether "nmstatectl format" reports any errors or how many.

Reply:
I rename the function to `yaml_correct`.

Reply:
But it seems to me that `nmstatectl format` does not report the number of errors, it just reports a error, if you fix this error and using `nmstatectl format` again, it will report another error if there is any errro again. 

```
[root@fcvm nmstate]# cat eth1_apply2.yml 
---                        
interfaces:                
  - name: eth1             
    type: ethernet       
    state: up              
    ipv4:                  
      address:             
        - ip: 192.168.3.2  
          prefix-length: 24
      dhcp: false          
      enabled: up        
    mtu: 1400i

[root@fcvm nmstate]# ./rust/target/debug/nmstatectl  -v format  eth1_apply2.yml
[2024-05-07T23:27:33Z INFO  nmstatectl] Nmstate version: 2.2.30
NmstateError: InvalidArgument: Invalid YAML string: interfaces: Need to be boolean: 1|0|true|false|yes|no|on|off|y|n at line 3 column 3
[root@fcvm nmstate]# vi eth1_apply2.yml 
[root@fcvm nmstate]# ./rust/target/debug/nmstatectl  -v format  eth1_apply2.yml
[2024-05-07T23:28:14Z INFO  nmstatectl] Nmstate version: 2.2.30
NmstateError: InvalidArgument: Invalid YAML string: interfaces: invalid digit found in string at line 3 column 3
```

Reply:
Can we change this in nmstatectl format?

Reply:
I checked with Gris, we can change this in `nmstatectl format`.

Reply:
I will discuss with Gris  to see if it is possible to report the number of errors.

Reply:
Currently, `nmstatectl format *.yml` only raises the first error it encounters, and the error is raised by the function, https://github.com/nmstate/nmstate/blob/v2.2.31/rust/src/lib/net_state.rs#L226

Various methods were tried and discussed with Gris: 

1. Recursively delete the lines that raise the errors, Gris thinks that recursively deleting the lines will cause
the syntax error if it has the subsections/subtree.
2. Patch serde_yaml. But serde_yaml is already archived and not maintained anymore, https://github.com/dtolnay/serde-yaml/releases/tag/0.9.34.
3. Enumerate each property in the YAML and validate each property in the function `validate_yaml_structure()`, but it is too cumbersome. 
https://github.com/liangwen12year/nmstate/commit/0b082806f4197755ef59eae255955ea6de1c48ef#diff-6a7ec70e192188c2b88a680d168c6fe16ca8c5d82fca08523c353029dfbad5a6R19


I discussed with Gris, it is not that easy and straightforward to support such functionality, the best possible approach is to patch serde_yaml and this method also needs a lot of effort regardless of whether serde_yaml is archived or maintained.

Reply:
Maybe the solution is to use a different YAML schema verifier, https://pypi.org/project/yamale/ seems to support reporting multiple errors at once.

Reply:
But I think that we still need to use `nmstatectl format *.yml` to serialize the settings in a proper order (normalized form).